### PR TITLE
feat: per branch cache+concurrency

### PR DESCRIPTION
**Pull request by innovaker**
Protects the cache from competing workflow runs on a per branch basis.

### CODE REVIEW

File: .github/workflows/containers.yml [hunk 1]
```diff
@@ -124,8 +128,8 @@ jobs:
             org.opencontainers.image.revision=${{ github.sha }}
           tags: |
             docker.io/${{ env.docker-hub-namespace }}/${{ steps.repository.outputs.name }}:${{ needs.tags.outputs.candidate }}
-          cache-from: type=registry,ref=docker.io/${{ env.docker-hub-namespace }}/${{ env.cache-repository-name }}:dev
-          cache-to: ${{ (steps.docker-hub-login.outcome == 'success') && (matrix.target == 'dev') && format('type=registry,ref=docker.io/{0}/{1}:{2},mode=max', env.docker-hub-namespace, env.cache-repository-name, 'dev') || null }}
+          cache-from: type=registry,ref=docker.io/${{ env.docker-hub-namespace }}/${{ env.cache-repository-name }}:${{ needs.tags.outputs.branch }}
+          cache-to: ${{ (steps.docker-hub-login.outcome == 'success') && (matrix.target == 'dev') && format('type=registry,ref=docker.io/{0}/{1}:{2},mode=max', env.docker-hub-namespace, env.cache-repository-name, needs.tags.outputs.branch) || null }}
```
Comment (petejohanson) [inline .github/workflows/containers.yml:132]: Do we want to in any way prefix/clarify this as being a caching image onlly, e.g.:

```suggestion
          cache-from: type=registry,ref=docker.io/${{ env.docker-hub-namespace }}/${{ env.cache-repository-name }}:cache-${{ needs.tags.outputs.branch }}
          cache-to: ${{ (steps.docker-hub-login.outcome == 'success') && (matrix.target == 'dev') && format('type=registry,ref=docker.io/{0}/{1}:cache-{2},mode=max', env.docker-hub-namespace, env.cache-repository-name, needs.tags.outputs.branch) || null }}
```

Comment (innovaker) [reply .github/workflows/containers.yml:132]: There's probably no need as the entire repository is solely for caching:
```
env:
  cache-repository-name: zmk-docker-cache
...
  ${{ env.cache-repository-name }}:${{ needs.tags.outputs.branch }}
```

Comment (innovaker) [reply .github/workflows/containers.yml:132]: The next PR will actually switch to `inline` cache (bundled with each image) which makes the cache repository redundant.  I very almost lumped it into this PR, but didn't want to confuse the issues.  The main purpose of this PR is make the switch from caching *per architecture* (which has already shown to be redundant) to *per branch* which allows each branch/PR to rebuild independently.

Comment (petejohanson) [reply .github/workflows/containers.yml:132]: Yeah, ok, missed that when reviewing this.

**Pull request merged**